File: python修改张铁海/p10.py
import numpy as np
import matplotlib.pyplot as plt
from mpl_toolkits.mplot3d import Axes3D

def getPositon(a):
    raw, column = a.shape  # get the matrix of a raw and column
    positon = np.argmax(a)  # get the index of max in the a
    # print(positon)
    r, c = divmod(positon , column)
    return r,c

def sence(p,z,world,pSenseCorrect):
    # nRow=len(p[0])
    # nCol=len(p[:0])
    nRow,nCol=np.shape(p)
    q_size=(nRow,nCol)
    q=np.zeros(q_size)
    for r in range(0,nRow):
        for c in range(0,nCol):
            if z== world[r][c]:
                hit=1
            else:
                hit=0
            q[r,c]=p[r,c]*(hit * pSenseCorrect + (1-hit) * (1-pSenseCorrect))
    S=sum(q)
    for i in range(len(q)):
        q[i]=q[i]/sum(S)
    return q

world = (('red', 'green', 'green', 'red', 'red'),
('red', 'red', 'green', 'red', 'red'),
('red', 'red', 'green', 'green', 'red'),
('red', 'red', 'red', 'red', 'red'))
nRow = len(world)
nCol = len(world[1])
print(world[1][1])
pStart = 0.7
ones_size=(nRow,nCol)
ones=np.ones(ones_size)
p = (1 - pStart) / (nRow * nCol - 1) * ones
pSenseCorrect=0.7
p[2,1]=pStart
measurements=['green']
# print(measurements[0])
q=sence(p,measurements[0],world,pSenseCorrect)
print("The Prior:\n",p)
print("The probability after sensing:\n",q)
r,c=getPositon(q)
print('The largest probability %s occurs at cell(%s,%s)\n'%(q[r,c],r,c))

plt.ion()
h=plt.figure(1)
ax=Axes3D(h)
ly= len(p[0])  #5            # Work out matrix dimensions
lx= len(p[:,0])  #4
xpos = np.arange(0,lx,1)    # Set up a mesh of positions
ypos = np.arange(0,ly,1)
xpos, ypos = np.meshgrid(xpos+0.25, ypos+0.25)
xpos = xpos.flatten()   # Convert positions to 1D array
ypos = ypos.flatten()
zpos = np.zeros(lx*ly)

dx = np.ones_like(zpos)
dy = dx.copy()
dz = p.T.flatten()
ax.bar3d(xpos,ypos,zpos, dx, dy, dz)

h1=plt.figure(2)
ax=Axes3D(h1)
ly= len(q[0])  #5            # Work out matrix dimensions
lx= len(q[:,0])  #4
xpos = np.arange(0,lx,1)    # Set up a mesh of positions
ypos = np.arange(0,ly,1)
xpos, ypos = np.meshgrid(xpos+0.25, ypos+0.25)
xpos = xpos.flatten()   # Convert positions to 1D array
ypos = ypos.flatten()
zpos = np.zeros(lx*ly)

dx = np.ones_like(zpos)
dy = dx.copy()
dz = q.T.flatten()
ax.bar3d(xpos,ypos,zpos, dx, dy, dz)
plt.ioff()
plt.show()

File: P10.cpp
// P10.cpp: 定义控制台应用程序的入口点。
//

#include "stdafx.h"
#include<iostream>
#include<string>
#include<iomanip>
const int nRow = 4;
const int nCol = 5;
const double pStart = 0.7;
const double pSenseCorrect = 0.7;

int main()
{
	using namespace std;
	double p[nRow][nCol];
	string Measurements[nRow][nCol];
	string world[nRow][nCol] = {{ "red","green","green","red","red" },
								{ "red","red","green","red","red" },
								{ "red","red","green","green","red" },
								{ "red","red","red","red","red" }, };
	double prior_value = (1 - pStart) / (nRow * nCol - 1);
	for (int i = 0; i < nRow; i++)
		for (int j = 0; j < nCol; j++)
			p[i][j] = prior_value;
	p[2][1] = pStart;
	for (int i = 0; i < nRow; i++)
		for (int  j = 0; j < nCol; j++)
			Measurements[i][j] = "green";
	int hit;
	double sum = 0;
	double q[nRow][nCol];
	for (int ii = 0; ii < nRow; ii++)
		for (int jj =0; jj< nCol; jj++)
		{
			if (Measurements[ii][jj] == world[ii][jj])
				hit = 1;
			else 
				hit = 0;
			q[ii][jj] = p[ii][jj] * (hit * pSenseCorrect + (1 - hit) * (1 - pSenseCorrect));
			sum = sum + q[ii][jj];
		}
	for (int i =0; i < nRow; i++)
		for (int j = 0; j < nCol; j++)
			q[i][j] = q[i][j] / sum;
	double q_max = q[1][1];
	int max_row =0; 
	int max_col = 0;
	for (int i = 0; i < nRow; i++)
		for (int j = 1; j < nCol; j++)
			if (q[i][j] > q_max)
			{
				q_max = q[i][j];
				max_row = i;
				max_col = j;
			}
	cout<<"The Prior: "<<endl;
	for (int i = 0; i < nRow; i++)
		for (int j = 0; j < nCol; j++)
		{
			cout << setiosflags(ios::fixed) << setprecision(7) << "p[" << i << "][" << j << "] : " << p[i][j] << "\t";			
			if (j == nCol)
				cout <<endl;
		}
	cout << endl << endl;
	cout<<"The probability after sensing : "<<endl;
	for (int i =0; i < nRow; i++)
		for (int j =0; j < nCol; j++)
		{
			cout << setiosflags(ios::fixed) << setprecision(7) << "q[" << i << "][" << j << "] : " << q[i][j] << "\t";
			if (j == nCol)
				cout << endl;
		}
	cout << endl;
	cout << "The largest probability is q[" << max_row << "]" << "[" << max_col << "] : " << q[max_row][max_col] << endl;
}


	
